src/ASM/ASMmxBase.C

```diff
@@ -352,6 +416,71 @@ ASMmxBase::VolumeVec ASMmxBase::establishBases(Go::SplineVolume* svol,
     result[3].reset(new Go::SplineVolume(*svol));
     geoBasis = 4;
+  } else if (type == SUBGRID) {
+    // basis1 should be one degree higher than basis2 and C^p-1 continuous
+    int ndim = svol->dimension();
+    result[1].reset(new Go::SplineVolume(*svol));
+    Go::SplineVolume tmp(*svol);
+    for (size_t dir = 0; dir < 3; ++dir) {
+      RealArray extraKnots;
+      RealArray::const_iterator uit = tmp.basis(dir).begin();
+      double ucurr, uprev = *(uit++);
+      while (uit != tmp.basis(dir).end())
+      {
+        ucurr = *(uit++);
+        if (ucurr > uprev)
+          for (int i = 0; i < 1; i++)
+          {
+            double xi = (double)(i+1)/(double)(1+1);
+            extraKnots.push_back(ucurr*xi + uprev*(1.0-xi));
+          }
+        uprev = ucurr;
+      }
+      tmp.insertKnot(dir, extraKnots);
+    }
+    Go::BsplineBasis b1 = tmp.basis(0).extendedBasis(tmp.order(0)+1);
+    Go::BsplineBasis b2 = tmp.basis(1).extendedBasis(tmp.order(1)+1);
+    Go::BsplineBasis b3 = tmp.basis(2).extendedBasis(tmp.order(2)+1);
+
+    // Compute parameter values of the Greville points
+    size_t i;
+    RealArray ug(b1.numCoefs()), vg(b2.numCoefs()), wg(b3.numCoefs());
+    for (i = 0; i < ug.size(); i++)
+      ug[i] = b1.grevilleParameter(i);
+    for (i = 0; i < vg.size(); i++)
+      vg[i] = b2.grevilleParameter(i);
+    for (i = 0; i < wg.size(); i++)
+      wg[i] = b3.grevilleParameter(i);
+
+    if (svol->rational()) {
+      std::vector<double> rCoefs(tmp.rcoefs_begin(), tmp.rcoefs_end());
+
+      // we normally would set coefs as (x*w, y*w, w)
+      // however, gotools use this representation internally already.
+
+      // instance a Bspline surface in ndim+1
+      Go::SplineVolume svol2(tmp.basis(0), tmp.basis(1), tmp.basis(2), rCoefs.begin(), ndim+1, false);
+
+      // interpolate the Bspline surface onto new basis
+      RealArray XYZ((ndim+1)*ug.size()*vg.size()*wg.size());
+      svol2.gridEvaluator(ug,vg,wg,XYZ);
+      std::unique_ptr<Go::SplineVolume> svol3(Go::VolumeInterpolator::regularInterpolation(b1,b2,b3,ug,vg,wg,XYZ,ndim+1,false,XYZ));
+
+      // new rational coefs are (x/w', y/w', w')
+      // apparently gotools will rescale coeffs on surface creation.
+      result[0].reset(new Go::SplineVolume(svol3->basis(0), svol3->basis(1), svol3->basis(2),svol3->coefs_begin(), ndim, true));
+    } else {
+      RealArray XYZ(ndim*ug.size()*vg.size()*wg.size());
+      // Evaluate the spline surface at all points
+      tmp.gridEvaluator(ug,vg,wg,XYZ);
+
+      // Project the coordinates onto the new basis (the 2nd XYZ is dummy here)
+      result[0].reset(Go::VolumeInterpolator::regularInterpolation(b1,b2,b3,
+                                                                   ug,vg,wg,
+                                                                   XYZ,ndim,
+                                                                   false,XYZ));
```

Comment:
Again: could not the interpolation be swapped out for knot insertion?

Reply:
as discussed on skype i do order raising along with the knot insertion so cannot comply.